src/cpp/session/modules/clang/CodeCompletion.cpp

```diff
@@ -154,6 +162,376 @@ core::json::Object toJson(const CodeCompleteResult& result)
+         if (!includePath.exists())
+            continue;
+         
+         pIncludePaths->push_back(includePath.absolutePath());
+      }
+   }
+}
+
+typedef std::pair<std::string, std::string> LinkingToEntry;
+std::vector<LinkingToEntry> parseLinkingTo(const std::string& linkingTo)
+{
+   static const boost::regex reLinkingTo(std::string() +
+         "^\\s*"       // initial whitespace
+         "(\\S*?)"     // package name
+         "\\s*"        // whitespace after
+         "(?:\\("      // begin optional capture group
+         "([^\\)]*?)"  // version (in parens)
+         "\\))?"       // closing paren + closing group
+         "\\s*$"       // ending whitespace
+   );
+         
+   std::vector<LinkingToEntry> result;
+   
+   // split on commas
+   std::vector<std::string> splat = core::algorithm::split(linkingTo, ",");
+   
+   // loop over our split pieces and extract packages
+   BOOST_FOREACH(const std::string& el, splat)
+   {
+      boost::smatch match;
+      if (boost::regex_match(el, match, reLinkingTo))
+      {
+         LinkingToEntry entry;
+         if (match.size() >= 2)
+            entry.first = match[1];
+         if (match.size() >= 3)
+            entry.second = match[2];
+         
+         result.push_back(entry);
+      }
+   }
+   
+   return result;
+}
+
+void discoverInstIncludePath(const FilePath& filePath,
+                             std::vector<std::string>* pIncludePaths)
+{
+   if (!filePath.exists())
+      return;
+   
+   FilePath pkgPath = findPackageRoot(filePath);
+   if (!pkgPath.exists())
+      return;
+   
+   // add 'inst/include path
+   FilePath instIncludePath = pkgPath.complete("inst/include");
+   if (instIncludePath.exists())
+      pIncludePaths->push_back(instIncludePath.absolutePath());
+   
+   // add paths to LinkingTo packages
+   std::string linkingTo = projects::projectContext().packageInfo().linkingTo();
+   std::vector<LinkingToEntry> entries = parseLinkingTo(linkingTo);
+   BOOST_FOREACH(const LinkingToEntry& entry, entries)
+   {
+      // attempt to resolve pkg path
+      FilePath pkgPath = findPackagePath(entry.first);
+      if (!pkgPath.exists())
+         continue;
+      
+      FilePath includePath = pkgPath.complete("include");
+      if (!includePath.exists())
+         continue;
+      
+      pIncludePaths->push_back(includePath.absolutePath());
+   }
+}
+
+Error getSystemHeaderCompletions(const std::string& token,
+                                 const std::string& parentDir,
+                                 const FilePath& filePath,
+                                 const std::string& docId,
+                                 const core::json::JsonRpcRequest& request,
+                                 core::json::JsonRpcResponse* pResponse)
+{
+   std::vector<std::string> includePaths;
+   
+   // discover the system headers
+   discoverSystemIncludePaths(&includePaths);
+   
+   // add R package include paths in library
```

Comment:
Beyond the core system headers (i.e. /usr/include) we actually already have code for discovering the include directories in play for a given translation unit (we actually need this for libclang to work at all). Better still this information is discovered by actually calling R CMD SHBLIB --dry-run and is cached. You can call `RCompilationDatabase::compileArgsForTranslationUnit` for a given source file and it will do the right thing if it's in a package (and otherwise treat it as a sourceCpp target). See: https://github.com/rstudio/rstudio/blob/master/src/cpp/session/modules/clang/RCompilationDatabase.cpp#L694-L695

In terms of what to do with the results of that function, there is an array of arguments to pass to the compiler and the ones prefaced with -I are the include directories. Note that this will also include the compilers built-in directories for e.g. the C++ standard library.

Assuming this mechanism works for this scenario, let's replace your ad-hoc directory discover code with a call to `compileArgsForTranslationUnit`